python/essex-config/essex_config/config.py

```diff
@@ -0,0 +1,297 @@
+            int,
+            ConfigurationField(),
+        ]
+
+
+    class TopLevelConfig(
+        Config
+    ):
+        sources: ClassVar[
+            list[Source]
+        ] = [
+            FileSource(
+                "config.yaml"
+            )
+        ]
+
+        sub_config: (
+            SubConfig
+        )
+        value1: Annotated[
+            str,
+            ConfigurationField(),
+        ]
+
+
+    config = TopLevelConfig.get_config()
+    assert (
+        config.value1
+        == "value1"
+    )
+    assert (
+        config.sub_config.value2
+        == 2
+    )
+    ```
+    """
+
+    sources: ClassVar[list[Source]] = DEFAULT_SOURCE_LIST
+
+    @staticmethod
+    def __resolve_value(names: list[str], data: dict[str, Any]) -> Any:
+        for name in names:
+            if name in data:
+                return data[name]
+
+            if "." in name:
+                parts = name.split(".")
+                value = data
+                try:
+                    for part in parts:
+                        value = value[part]
+                except KeyError:
+                    continue
+                return value
+
+        return None
+
+    @staticmethod
+    @cache
+    def __get_data_from_source(source: Source) -> dict[str, Any]:
+        return source.get_data()
+
+    @classmethod
+    def get_config(
+        cls: type[T], refresh_config: bool = False, parents: str | None = None
+    ) -> T:
+        """Instantiate the configuration and all values.
+
+        Creates an instance of the Configuration class based on the values of data. If data is None, it will use the environment variables.
+
+        Parameters
+        ----------
+            refresh_config : bool, optional
+                If True, it will refresh the configuration values, by default False
+            parents: str, optional
+                The parent class name, used for nested configurations, by default None
+
+        Returns
+        -------
+            Config: Instance of the configuration class.
+
+        Raises
+        ------
+            ValueError
+                If any of the values is not found.
+        """
+        parents = cls.__name__ if parents is None else f"{parents}.{cls.__name__}"
+
+        if refresh_config:
+            cls.__get_data_from_source.cache_clear()
+
+        values = {}
+        for name, info in cls.model_fields.items():
+            if issubclass(
+                cast(type, info.annotation), Config
+            ):  # If the field is a subclass of Config then create an instance and assign it
+                sub_config_cls: type[Config] = cast(type[Config], info.annotation)
+                values[name] = sub_config_cls.get_config(parents=parents)
+                continue
+
+            if len(info.metadata) == 0 or not isinstance(
+                info.metadata[0], ConfigurationField
+            ):  # If there's a field without ConfigurationField then ignore it
```

Comment:
Should we ignore it? Or should we look for a match on the target anyway?

I would expect to be able to not specify `ConfigurationField()` when I don't want to change the default behavior.

Reply:
I changed the behavior so that we don't ignore it anymore, I don't really know why I added this